In [49]:
## import the dependencies
## dependencies are important because instead of writing a bunch of code to do a certain task from scratch, someone
###else or another company already created it for you to use

import json
import pandas as pd
import numpy as np
import re ## going to need this for regular expressions

In [2]:
## create the file path to your json file
## its currently saved in the module 9 folder inside surfs_up

file_dir = 'wikipedia-movies.json'

In [3]:
## we don't want to load messy json data into a df, so we need to create a list of dictionaries

with open(file_dir, mode='r') as file:
    wiki_movies_raw = json.load(file)

In [4]:
len(wiki_movies_raw)

7311

In [5]:
## see the first 5 records
## square bracket is a list

wiki_movies_raw[:5]

[{'url': 'https://en.wikipedia.org/wiki/The_Adventures_of_Ford_Fairlane',
  'year': 1990,
  'imdb_link': 'https://www.imdb.com/title/tt0098987/',
  'title': 'The Adventures of Ford Fairlane',
  'Directed by': 'Renny Harlin',
  'Produced by': ['Steve Perry', 'Joel Silver'],
  'Screenplay by': ['David Arnott', 'James Cappe', 'Daniel Waters'],
  'Story by': ['David Arnott', 'James Cappe'],
  'Based on': ['Characters', 'by Rex Weiner'],
  'Starring': ['Andrew Dice Clay',
   'Wayne Newton',
   'Priscilla Presley',
   'Lauren Holly',
   'Morris Day',
   'Robert Englund',
   "Ed O'Neill"],
  'Narrated by': 'Andrew "Dice" Clay',
  'Music by': ['Cliff Eidelman', 'Yello'],
  'Cinematography': 'Oliver Wood',
  'Edited by': 'Michael Tronick',
  'Productioncompany ': 'Silver Pictures',
  'Distributed by': '20th Century Fox',
  'Release date': ['July 11, 1990', '(', '1990-07-11', ')'],
  'Running time': '102 minutes',
  'Country': 'United States',
  'Language': 'English',
  'Budget': '$20 million',


In [6]:
## see the last 5 records by using negative index slices
wiki_movies_raw[-5]

{'url': 'https://en.wikipedia.org/wiki/Holmes_%26_Watson',
 'year': 2018,
 'imdb_link': 'https://www.imdb.com/title/tt1255919/',
 'title': 'Holmes & Watson',
 'Directed by': 'Etan Cohen',
 'Produced by': ['Will Ferrell',
  'Adam McKay',
  'Jimmy Miller',
  'Clayton Townsend'],
 'Screenplay by': 'Etan Cohen',
 'Based on': ['Sherlock Holmes',
  'and',
  'Dr. Watson',
  'by',
  'Sir Arthur Conan Doyle'],
 'Starring': ['Will Ferrell',
  'John C. Reilly',
  'Rebecca Hall',
  'Rob Brydon',
  'Steve Coogan',
  'Ralph Fiennes'],
 'Music by': 'Mark Mothersbaugh',
 'Cinematography': 'Oliver Wood',
 'Edited by': 'Dean Zimmerman',
 'Productioncompanies ': ['Columbia Pictures',
  'Gary Sanchez Productions',
  'Mosaic Media Group',
  'Mimran Schur Pictures'],
 'Distributed by': 'Sony Pictures Releasing',
 'Release date': ['December 25, 2018',
  '(',
  '2018-12-25',
  ')',
  '(United States)'],
 'Running time': '90 minutes',
 'Country': 'United States',
 'Language': 'English',
 'Budget': '$42 million

In [7]:
## checking random records in the middle
wiki_movies_raw[3600:3605]

[{'url': 'https://en.wikipedia.org/wiki/Benji:_Off_the_Leash!',
  'year': 2004,
  'imdb_link': 'https://www.imdb.com/title/tt0315273/',
  'title': 'Benji: Off the Leash!',
  'Directed by': 'Joe Camp',
  'Written by': 'Joe Camp',
  'Starring': ['Benji', 'Nick Whitaker', 'Shaggy', 'Gypsy the Cockatoo'],
  'Music by': 'Antonio di Lorenzo',
  'Productioncompany ': 'Mulberry Square Productions',
  'Distributed by': 'Mulberry Square Productions',
  'Release date': ['March 26, 2004', '(', '2004-03-26', ')'],
  'Running time': '97 min',
  'Country': 'United States',
  'Language': 'English',
  'Box office': '$3,817,362'},
 {'url': 'https://en.wikipedia.org/wiki/The_Best_Thief_in_the_World',
  'year': 2004,
  'imdb_link': 'https://www.imdb.com/title/tt0389796/',
  'title': 'The Best Thief in the World',
  'Directed by': 'Jacob Kornbluth',
  'Produced by': ['Tim Perrell', 'Nicola Usborne'],
  'Written by': 'Jacob Kornbluth',
  'Starring': ['Marc Rozendaal',
   'Michael Silverman',
   'David Warsh

In [8]:
## kaggle data can be imported directly to df because they are flat-files (csv)
kaggle_metadata = pd.read_csv('movies_metadata.csv', low_memory=False)
ratings = pd.read_csv('ratings.csv')

In [9]:
## inspecting the data
kaggle_metadata.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [10]:
ratings.head()

userId  movieId  rating   timestamp
0       1      110     1.0  1425941529
1       1      147     4.5  1425942435
2       1      858     5.0  1425941523
3       1     1221     5.0  1425941546
4       1     1246     5.0  1425941556

### initial investigation
* easiest way to find errors, just pretend there aren't any and import data into a dataframe

In [11]:
## convert the wiki_movies_raw data to a dataframe
wiki_movies_df = pd.DataFrame(wiki_movies_raw)

In [12]:
wiki_movies_df.head()

url    year  \
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990.0   
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990.0   
2   https://en.wikipedia.org/wiki/Air_America_(film)  1990.0   
3    https://en.wikipedia.org/wiki/Alice_(1990_film)  1990.0   
4      https://en.wikipedia.org/wiki/Almost_an_Angel  1990.0   

                               imdb_link                            title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   
2  https://www.imdb.com/title/tt0099005/                      Air America   
3  https://www.imdb.com/title/tt0099012/                            Alice   
4  https://www.imdb.com/title/tt0099018/                  Almost an Angel   

          Directed by                  Produced by  \
0        Renny Harlin   [Steve Perry, Joel Silver]   
1         James Foley  [Ric Kidney, Robert Redlin]   
2  Roger Spottiswoode               Daniel Melnick   
3         Woody Allen              Robert Greenhut   
4        John Cornell                 John Cornell   

                                Screenplay by                     Story by  \
0  [David Arnott, James Cappe, Daniel Waters]  [David Arnott, James Cappe]   
1                [James Foley, Robert Redlin]                          NaN   
2                  [John Eskow, Richard Rush]                          NaN   
3                                         NaN                          NaN   
4                                         NaN                          NaN   

                                            Based on  \
0                        [Characters, by Rex Weiner]   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   
2             [Air America, by, Christopher Robbins]   
3                                                NaN   
4                                                NaN   

                                            Starring  ... Predecessor  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...  ...         NaN   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...  ...         NaN   
2  [Mel Gibson, Robert Downey Jr., Nancy Travis, ...  ...         NaN   
3  [Alec Baldwin, Blythe Danner, Judy Davis, Mia ...  ...         NaN   
4        [Paul Hogan, Elias Koteas, Linda Kozlowski]  ...         NaN   

  Founders Area served Products Services Russian Hebrew Revenue  \
0      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
1      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
2      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
3      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
4      NaN         NaN      NaN      NaN     NaN    NaN     NaN   

  Operating income Polish  
0              NaN    NaN  
1              NaN    NaN  
2              NaN    NaN  
3              NaN    NaN  
4              NaN    NaN  

[5 rows x 193 columns]

### use the column count 
* using the row and column count, you can see there are 5 rows (expected because we're using head) but there are 193 columns

In [13]:
## convert wiki_movies_df columns to a list to see the names of all the columns
wiki_movies_df.columns.tolist()

['url',
 'year',
 'imdb_link',
 'title',
 'Directed by',
 'Produced by',
 'Screenplay by',
 'Story by',
 'Based on',
 'Starring',
 'Narrated by',
 'Music by',
 'Cinematography',
 'Edited by',
 'Productioncompany ',
 'Distributed by',
 'Release date',
 'Running time',
 'Country',
 'Language',
 'Budget',
 'Box office',
 'Written by',
 'Genre',
 'Theme music composer',
 'Country of origin',
 'Original language(s)',
 'Producer(s)',
 'Editor(s)',
 'Production company(s)',
 'Original network',
 'Original release',
 'Productioncompanies ',
 'Executive producer(s)',
 'Production location(s)',
 'Distributor',
 'Picture format',
 'Audio format',
 'Voices of',
 'Followed by',
 'Composer(s)',
 'Created by',
 'Also known as',
 'Opening theme',
 'No. of episodes',
 'Preceded by',
 'Author',
 'Publisher',
 'Publication date',
 'Media type',
 'Pages',
 'ISBN',
 'OCLC',
 'LC Class',
 'Cover artist',
 'Series',
 'Set in',
 'Adaptation by',
 'Suggested by',
 'Biographical data',
 'Born',
 'Died',
 'Resti

In [14]:
## need to modify the json data by redistricting it to only those entries that have a director and an imdb link. 
## can be done with list comprehension -- use to filter data

wiki_movies = [movie for movie in wiki_movies_raw
               if ('Director' in movie or 'Directed by' in movie)
                   and 'imdb_link' in movie]
len(wiki_movies)

7080

In [15]:
## make a dataframe from wiki_movies
wiki_movies_df = pd.DataFrame(wiki_movies)

In [16]:
wiki_movies_df.head()

url  year  \
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990   
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990   
2   https://en.wikipedia.org/wiki/Air_America_(film)  1990   
3    https://en.wikipedia.org/wiki/Alice_(1990_film)  1990   
4      https://en.wikipedia.org/wiki/Almost_an_Angel  1990   

                               imdb_link                            title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   
2  https://www.imdb.com/title/tt0099005/                      Air America   
3  https://www.imdb.com/title/tt0099012/                            Alice   
4  https://www.imdb.com/title/tt0099018/                  Almost an Angel   

          Directed by                  Produced by  \
0        Renny Harlin   [Steve Perry, Joel Silver]   
1         James Foley  [Ric Kidney, Robert Redlin]   
2  Roger Spottiswoode               Daniel Melnick   
3         Woody Allen              Robert Greenhut   
4        John Cornell                 John Cornell   

                                Screenplay by                     Story by  \
0  [David Arnott, James Cappe, Daniel Waters]  [David Arnott, James Cappe]   
1                [James Foley, Robert Redlin]                          NaN   
2                  [John Eskow, Richard Rush]                          NaN   
3                                         NaN                          NaN   
4                                         NaN                          NaN   

                                            Based on  \
0                        [Characters, by Rex Weiner]   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   
2             [Air America, by, Christopher Robbins]   
3                                                NaN   
4                                                NaN   

                                            Starring  ... Hepburn Literally  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...  ...     NaN       NaN   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...  ...     NaN       NaN   
2  [Mel Gibson, Robert Downey Jr., Nancy Travis, ...  ...     NaN       NaN   
3  [Alec Baldwin, Blythe Danner, Judy Davis, Mia ...  ...     NaN       NaN   
4        [Paul Hogan, Elias Koteas, Linda Kozlowski]  ...     NaN       NaN   

  Cantonese Chinese Yiddish Arabic Romanized Russian Hebrew Polish  
0       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
1       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
2       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
3       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
4       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  

[5 rows x 78 columns]

In [17]:
## look through the edited column list again
wiki_movies_df.columns.tolist()

['url',
 'year',
 'imdb_link',
 'title',
 'Directed by',
 'Produced by',
 'Screenplay by',
 'Story by',
 'Based on',
 'Starring',
 'Narrated by',
 'Music by',
 'Cinematography',
 'Edited by',
 'Productioncompany ',
 'Distributed by',
 'Release date',
 'Running time',
 'Country',
 'Language',
 'Budget',
 'Box office',
 'Written by',
 'Genre',
 'Theme music composer',
 'Country of origin',
 'Original language(s)',
 'Producer(s)',
 'Editor(s)',
 'Production company(s)',
 'Original network',
 'Original release',
 'Productioncompanies ',
 'Executive producer(s)',
 'Production location(s)',
 'Distributor',
 'Picture format',
 'Audio format',
 'Voices of',
 'Followed by',
 'Composer(s)',
 'Created by',
 'Also known as',
 'Opening theme',
 'No. of episodes',
 'Preceded by',
 'Adaptation by',
 'Suggested by',
 'Traditional',
 'Mandarin',
 'Released',
 'Recorded',
 'Venue',
 'Length',
 'Label',
 'Director',
 'Producer',
 'Animation by',
 'Color process',
 'Simplified',
 'French',
 'Camera setup'

In [18]:
## looking through the list of columns, there are a lot of languages and there is also one called
## no. of episodes -- this means there are some that are not movies and may just be a television show
## create a list comprehension to FILTER the data

wiki_movies = [movie for movie in wiki_movies_raw
               if ('Director' in movie or 'Directed by' in movie)
                  and 'imdb_link' in movie
                  and 'No. of episodes' not in movie]
wiki_movies_df = pd.DataFrame(wiki_movies)

In [19]:
wiki_movies_df.head()

url  year  \
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990   
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990   
2   https://en.wikipedia.org/wiki/Air_America_(film)  1990   
3    https://en.wikipedia.org/wiki/Alice_(1990_film)  1990   
4      https://en.wikipedia.org/wiki/Almost_an_Angel  1990   

                               imdb_link                            title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   
2  https://www.imdb.com/title/tt0099005/                      Air America   
3  https://www.imdb.com/title/tt0099012/                            Alice   
4  https://www.imdb.com/title/tt0099018/                  Almost an Angel   

          Directed by                  Produced by  \
0        Renny Harlin   [Steve Perry, Joel Silver]   
1         James Foley  [Ric Kidney, Robert Redlin]   
2  Roger Spottiswoode               Daniel Melnick   
3         Woody Allen              Robert Greenhut   
4        John Cornell                 John Cornell   

                                Screenplay by                     Story by  \
0  [David Arnott, James Cappe, Daniel Waters]  [David Arnott, James Cappe]   
1                [James Foley, Robert Redlin]                          NaN   
2                  [John Eskow, Richard Rush]                          NaN   
3                                         NaN                          NaN   
4                                         NaN                          NaN   

                                            Based on  \
0                        [Characters, by Rex Weiner]   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   
2             [Air America, by, Christopher Robbins]   
3                                                NaN   
4                                                NaN   

                                            Starring  ... Hepburn Literally  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...  ...     NaN       NaN   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...  ...     NaN       NaN   
2  [Mel Gibson, Robert Downey Jr., Nancy Travis, ...  ...     NaN       NaN   
3  [Alec Baldwin, Blythe Danner, Judy Davis, Mia ...  ...     NaN       NaN   
4        [Paul Hogan, Elias Koteas, Linda Kozlowski]  ...     NaN       NaN   

  Cantonese Chinese Yiddish Arabic Romanized Russian Hebrew Polish  
0       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
1       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
2       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
3       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
4       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  

[5 rows x 75 columns]

### practicing scope of variables
* we can create a local variable (created inside the function) with the same name as a global variable (created outside function)

In [20]:
x = 'global value' ### this variable was created outside the function

def foo(): ### def means you're defining the function
    x = 'local value' ### this is a local variable but you still named it x
    print(x) ### print the variable
    
foo() ### calling the defined function (i think)
print(x) ### print the global value bc its outside

local value
global value


In [21]:
my_list = [1,2,3] ## creating the global variable bc it was created outside the function

def append_four(x): ## defining the function -- add 4 to the end of the list
    x.append(4) 
append_four(my_list)
print(my_list) 

[1, 2, 3, 4]


In [22]:
## this changed the 'my_list' variable to include the 4
## if we want to leave the original object (my_list) unchanged, then we need to make a copy

### new_list = list(old_list)
### new_dict = dict(old_dict)

### lambda function example

In [23]:
## how to square a value
square = lambda x: x * x ## (x) is the arguement (x * x) is the expression 'square' is the name
square(5) ### we're calling our named function and assigning 5 as the x in our formula

25

In [24]:
 ## we start the function by using def
def clean_movie(movie):
    movie = dict(movie) ## create a non-destructive copy. we're assigning movie to a local (inside) variable. 
    return movie

In [25]:
## call all the movies that have something written in the arabic column and is not null
wiki_movies_df[wiki_movies_df['Arabic'].notnull()]

url  year  \
6834  https://en.wikipedia.org/wiki/The_Insult_(film)  2018   
7058   https://en.wikipedia.org/wiki/Capernaum_(film)  2018   

                                  imdb_link       title    Directed by  \
6834  https://www.imdb.com/title/tt7048622/  The Insult   Ziad Doueiri   
7058  https://www.imdb.com/title/tt8267604/   Capernaum  Nadine Labaki   

                                            Produced by  \
6834  [Rachid Bouchareb, Jean Bréhat, Julie Gayet, A...   
7058                    [Michel Merkt, Khaled Mouzanar]   

                                          Screenplay by  \
6834                                                NaN   
7058  [Nadine Labaki, Jihad Hojaily, Michelle Keserw...   

                                               Story by Based on  \
6834                                                NaN      NaN   
7058  [Georges Khabbaz, Nadine Labaki, Michelle Kese...      NaN   

                                               Starring  ... Hepburn  \
6834                       [Adel Karam, Kamel El Basha]  ...     NaN   
7058  [Zain Al Rafeea, Yordanos Shiferaw, Boluwatife...  ...     NaN   

        Literally Cantonese Chinese Yiddish       Arabic        Romanized  \
6834  Case No. 23       NaN     NaN     NaN  قضية رقم ٢٣  Qadiyya raqm 23   
7058          NaN       NaN     NaN     NaN    کفرناحوم‎              NaN   

     Russian Hebrew Polish  
6834     NaN    NaN    NaN  
7058     NaN    NaN    NaN  

[2 rows x 75 columns]

In [26]:
wiki_movies_df[wiki_movies_df['Arabic'].notnull()]['url'] ## pull the url

6834    https://en.wikipedia.org/wiki/The_Insult_(film)
7058     https://en.wikipedia.org/wiki/Capernaum_(film)
Name: url, dtype: object

In [27]:
## sort in alphabetical order

sorted(wiki_movies_df.columns.tolist())

['Adaptation by',
 'Also known as',
 'Animation by',
 'Arabic',
 'Audio format',
 'Based on',
 'Box office',
 'Budget',
 'Cantonese',
 'Chinese',
 'Cinematography',
 'Color process',
 'Composer(s)',
 'Country',
 'Country of origin',
 'Created by',
 'Directed by',
 'Director',
 'Distributed by',
 'Distributor',
 'Edited by',
 'Editor(s)',
 'Executive producer(s)',
 'Followed by',
 'French',
 'Genre',
 'Hangul',
 'Hebrew',
 'Hepburn',
 'Japanese',
 'Label',
 'Language',
 'Length',
 'Literally',
 'Mandarin',
 'McCune–Reischauer',
 'Music by',
 'Narrated by',
 'Original language(s)',
 'Original network',
 'Original release',
 'Original title',
 'Picture format',
 'Polish',
 'Preceded by',
 'Produced by',
 'Producer',
 'Producer(s)',
 'Production company(s)',
 'Production location(s)',
 'Productioncompanies ',
 'Productioncompany ',
 'Recorded',
 'Release date',
 'Released',
 'Revised Romanization',
 'Romanized',
 'Running time',
 'Russian',
 'Screen story by',
 'Screenplay by',
 'Simplifie

In [28]:
wiki_movies_df['Polish'].value_counts() ### the language columns hold the alternate titles for the movies

Zimna wojna    1
Name: Polish, dtype: int64

### handling the alternative titles
Logic
* make an empty dict to hold all the alternative keys
* loop through a list of all alternative title keys
- check if the current key exists in the movie object
- if so, remove the key-value pair and add to the alternative titles dict
* after looping through every key, add the alternative titles dict to the movie object

In [29]:
## step 1: make an empty dict to hold all of the alternative titles
def clean_movie(movie):
    movie = dict(movie) # this is to create a non-destructive copy
    alt_titles = {} # empty dict -- curly brackets, empty list -- square brackets
    for key in ['Also known as','Arabic','Cantonese','Chinese','French', ## step2: loop through a list of all alt title keys
                'Hangul','Hebrew','Hepburn','Japanese','Literally',
                'Mandarin','McCune–Reischauer','Original title','Polish',
                'Revised Romanization','Romanized','Russian',
                'Simplified','Traditional','Yiddish']: ## creating another loop
        if key in movie: ### if so, remove the key-value pair and add to the alt titles dict
            alt_titles[key] = movie[key]
            movie.pop(key)
    if len(alt_titles) > 0: ## after looping through every key, add the alt titles to movie object
        movie['alt_titles'] = alt_titles
    return movie

In [30]:
## create a list of cleaned movies with a list comprehension
clean_movies = [clean_movie(movie) for movie in wiki_movies]

In [31]:
## set wiki_movies_df to be the df created from clean_movies
wiki_movies_df = pd.DataFrame(clean_movies)
sorted(wiki_movies_df.columns.tolist())

['Adaptation by',
 'Animation by',
 'Audio format',
 'Based on',
 'Box office',
 'Budget',
 'Cinematography',
 'Color process',
 'Composer(s)',
 'Country',
 'Country of origin',
 'Created by',
 'Directed by',
 'Director',
 'Distributed by',
 'Distributor',
 'Edited by',
 'Editor(s)',
 'Executive producer(s)',
 'Followed by',
 'Genre',
 'Label',
 'Language',
 'Length',
 'Music by',
 'Narrated by',
 'Original language(s)',
 'Original network',
 'Original release',
 'Picture format',
 'Preceded by',
 'Produced by',
 'Producer',
 'Producer(s)',
 'Production company(s)',
 'Production location(s)',
 'Productioncompanies ',
 'Productioncompany ',
 'Recorded',
 'Release date',
 'Released',
 'Running time',
 'Screen story by',
 'Screenplay by',
 'Starring',
 'Story by',
 'Suggested by',
 'Theme music composer',
 'Venue',
 'Voices of',
 'Written by',
 'alt_titles',
 'imdb_link',
 'title',
 'url',
 'year']

In [32]:
def clean_movie(movie):
    movie = dict(movie) #create a non-destructive copy
    alt_titles = {}
    # combine alternate titles into one list
    for key in ['Also known as','Arabic','Cantonese','Chinese','French',
                'Hangul','Hebrew','Hepburn','Japanese','Literally',
                'Mandarin','McCune-Reischauer','Original title','Polish',
                'Revised Romanization','Romanized','Russian',
                'Simplified','Traditional','Yiddish']:
        if key in movie:
            alt_titles[key] = movie[key]
            movie.pop(key)
    if len(alt_titles) > 0:
        movie['alt_titles'] = alt_titles

    # merge column names
    def change_column_name(old_name, new_name):
        if old_name in movie:
            movie[new_name] = movie.pop(old_name)
    change_column_name('Adaptation by', 'Writer(s)')
    change_column_name('Country of origin', 'Country')
    change_column_name('Directed by', 'Director')
    change_column_name('Distributed by', 'Distributor')
    change_column_name('Edited by', 'Editor(s)')
    change_column_name('Length', 'Running time')
    change_column_name('Original release', 'Release date')
    change_column_name('Music by', 'Composer(s)')
    change_column_name('Produced by', 'Producer(s)')
    change_column_name('Producer', 'Producer(s)')
    change_column_name('Productioncompanies ', 'Production company(s)')
    change_column_name('Productioncompany ', 'Production company(s)')
    change_column_name('Released', 'Release Date')
    change_column_name('Release Date', 'Release date')
    change_column_name('Screen story by', 'Writer(s)')
    change_column_name('Screenplay by', 'Writer(s)')
    change_column_name('Story by', 'Writer(s)')
    change_column_name('Theme music composer', 'Composer(s)')
    change_column_name('Written by', 'Writer(s)')

    return movie

In [33]:
clean_movies = [clean_movie(movie) for movie in wiki_movies]
wiki_movies_df = pd.DataFrame(clean_movies)
sorted(wiki_movies_df.columns.tolist())

['Animation by',
 'Audio format',
 'Based on',
 'Box office',
 'Budget',
 'Cinematography',
 'Color process',
 'Composer(s)',
 'Country',
 'Created by',
 'Director',
 'Distributor',
 'Editor(s)',
 'Executive producer(s)',
 'Followed by',
 'Genre',
 'Label',
 'Language',
 'McCune–Reischauer',
 'Narrated by',
 'Original language(s)',
 'Original network',
 'Picture format',
 'Preceded by',
 'Producer(s)',
 'Production company(s)',
 'Production location(s)',
 'Recorded',
 'Release date',
 'Running time',
 'Starring',
 'Suggested by',
 'Venue',
 'Voices of',
 'Writer(s)',
 'alt_titles',
 'imdb_link',
 'title',
 'url',
 'year']

In [34]:
## how to find the imbd id tt1234567 
## () look for one group
## match the tt string
## \d match numerical digit
## {7} match the last thing exactly 7 times

wiki_movies_df['imdb_id'] = wiki_movies_df['imdb_link'].str.extract(r'(tt\d{7})')
print(len(wiki_movies_df))

7076


In [35]:
## use drop duplicates to drop the duplicates
wiki_movies_df.drop_duplicates(subset='imdb_id', inplace=True)
print(len(wiki_movies_df))
wiki_movies_df.head()

7033


url  year  \
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990   
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990   
2   https://en.wikipedia.org/wiki/Air_America_(film)  1990   
3    https://en.wikipedia.org/wiki/Alice_(1990_film)  1990   
4      https://en.wikipedia.org/wiki/Almost_an_Angel  1990   

                               imdb_link                            title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   
2  https://www.imdb.com/title/tt0099005/                      Air America   
3  https://www.imdb.com/title/tt0099012/                            Alice   
4  https://www.imdb.com/title/tt0099018/                  Almost an Angel   

                                            Based on  \
0                        [Characters, by Rex Weiner]   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   
2             [Air America, by, Christopher Robbins]   
3                                                NaN   
4                                                NaN   

                                            Starring         Narrated by  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...  Andrew "Dice" Clay   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...                 NaN   
2  [Mel Gibson, Robert Downey Jr., Nancy Travis, ...                 NaN   
3  [Alec Baldwin, Blythe Danner, Judy Davis, Mia ...                 NaN   
4        [Paul Hogan, Elias Koteas, Linda Kozlowski]                 NaN   

   Cinematography                                       Release date  \
0     Oliver Wood                  [July 11, 1990, (, 1990-07-11, )]   
1    Mark Plummer  [May 17, 1990, (, 1990-05-17, ), (Cannes Film ...   
2   Roger Deakins                [August 10, 1990, (, 1990-08-10, )]   
3  Carlo Di Palma              [December 25, 1990, (, 1990-12-25, )]   
4    Russell Boyd                                  December 19, 1990   

  Running time  ... Preceded by Suggested by alt_titles Recorded Venue Label  \
0  102 minutes  ...         NaN          NaN        NaN      NaN   NaN   NaN   
1  114 minutes  ...         NaN          NaN        NaN      NaN   NaN   NaN   
2  113 minutes  ...         NaN          NaN        NaN      NaN   NaN   NaN   
3  106 minutes  ...         NaN          NaN        NaN      NaN   NaN   NaN   
4   95 minutes  ...         NaN          NaN        NaN      NaN   NaN   NaN   

  Animation by Color process McCune–Reischauer    imdb_id  
0          NaN           NaN               NaN  tt0098987  
1          NaN           NaN               NaN  tt0098994  
2          NaN           NaN               NaN  tt0099005  
3          NaN           NaN               NaN  tt0099012  
4          NaN           NaN               NaN  tt0099018  

[5 rows x 41 columns]

In [36]:
## use a list comprehension to get the count of null values
[[column,wiki_movies_df[column].isnull().sum()] for column in wiki_movies_df.columns]

[['url', 0],
 ['year', 0],
 ['imdb_link', 0],
 ['title', 1],
 ['Based on', 4852],
 ['Starring', 184],
 ['Narrated by', 6752],
 ['Cinematography', 691],
 ['Release date', 32],
 ['Running time', 139],
 ['Country', 236],
 ['Language', 244],
 ['Budget', 2295],
 ['Box office', 1548],
 ['Director', 0],
 ['Distributor', 357],
 ['Editor(s)', 548],
 ['Composer(s)', 518],
 ['Producer(s)', 202],
 ['Production company(s)', 1678],
 ['Writer(s)', 199],
 ['Genre', 6923],
 ['Original language(s)', 6875],
 ['Original network', 6908],
 ['Executive producer(s)', 6936],
 ['Production location(s)', 6986],
 ['Picture format', 6969],
 ['Audio format', 6972],
 ['Voices of', 7031],
 ['Followed by', 7024],
 ['Created by', 7023],
 ['Preceded by', 7023],
 ['Suggested by', 7032],
 ['alt_titles', 7012],
 ['Recorded', 7031],
 ['Venue', 7032],
 ['Label', 7031],
 ['Animation by', 7031],
 ['Color process', 7032],
 ['McCune–Reischauer', 7031],
 ['imdb_id', 0]]

In [38]:
wiki_columns_to_keep = [column for column in wiki_movies_df.columns if wiki_movies_df[column].isnull().sum() < len(wiki_movies_df) * 0.9]
wiki_movies_df = wiki_movies_df[wiki_columns_to_keep]
wiki_movies_df.head()

url  year  \
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990   
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990   
2   https://en.wikipedia.org/wiki/Air_America_(film)  1990   
3    https://en.wikipedia.org/wiki/Alice_(1990_film)  1990   
4      https://en.wikipedia.org/wiki/Almost_an_Angel  1990   

                               imdb_link                            title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   
2  https://www.imdb.com/title/tt0099005/                      Air America   
3  https://www.imdb.com/title/tt0099012/                            Alice   
4  https://www.imdb.com/title/tt0099018/                  Almost an Angel   

                                            Based on  \
0                        [Characters, by Rex Weiner]   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   
2             [Air America, by, Christopher Robbins]   
3                                                NaN   
4                                                NaN   

                                            Starring  Cinematography  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...     Oliver Wood   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...    Mark Plummer   
2  [Mel Gibson, Robert Downey Jr., Nancy Travis, ...   Roger Deakins   
3  [Alec Baldwin, Blythe Danner, Judy Davis, Mia ...  Carlo Di Palma   
4        [Paul Hogan, Elias Koteas, Linda Kozlowski]    Russell Boyd   

                                        Release date Running time  \
0                  [July 11, 1990, (, 1990-07-11, )]  102 minutes   
1  [May 17, 1990, (, 1990-05-17, ), (Cannes Film ...  114 minutes   
2                [August 10, 1990, (, 1990-08-10, )]  113 minutes   
3              [December 25, 1990, (, 1990-12-25, )]  106 minutes   
4                                  December 19, 1990   95 minutes   

         Country  ...       Budget        Box office            Director  \
0  United States  ...  $20 million     $21.4 million        Renny Harlin   
1  United States  ...   $6 million      $2.7 million         James Foley   
2  United States  ...  $35 million       $57,718,089  Roger Spottiswoode   
3  United States  ...  $12 million        $7,331,647         Woody Allen   
4             US  ...  $25 million  $6,939,946 (USA)        John Cornell   

          Distributor                       Editor(s)  \
0    20th Century Fox                 Michael Tronick   
1     Avenue Pictures                 Howard E. Smith   
2    TriStar Pictures  [John Bloom, Lois Freeman-Fox]   
3      Orion Pictures                  Susan E. Morse   
4  Paramount Pictures                    David Stiven   

               Composer(s)                  Producer(s)  \
0  [Cliff Eidelman, Yello]   [Steve Perry, Joel Silver]   
1            Maurice Jarre  [Ric Kidney, Robert Redlin]   
2            Charles Gross               Daniel Melnick   
3                      NaN              Robert Greenhut   
4            Maurice Jarre                 John Cornell   

                   Production company(s)                     Writer(s)  \
0                        Silver Pictures   [David Arnott, James Cappe]   
1                        Avenue Pictures  [James Foley, Robert Redlin]   
2  [Carolco Pictures, IndieProd Company]    [John Eskow, Richard Rush]   
3                                    NaN                   Woody Allen   
4                                    NaN                    Paul Hogan   

     imdb_id  
0  tt0098987  
1  tt0098994  
2  tt0099005  
3  tt0099012  
4  tt0099018  

[5 rows x 21 columns]

In [39]:
## displays the data type for each column
wiki_movies_df.dtypes

url                      object
year                      int64
imdb_link                object
title                    object
Based on                 object
Starring                 object
Cinematography           object
Release date             object
Running time             object
Country                  object
Language                 object
Budget                   object
Box office               object
Director                 object
Distributor              object
Editor(s)                object
Composer(s)              object
Producer(s)              object
Production company(s)    object
Writer(s)                object
imdb_id                  object
dtype: object

In [43]:
box_office = wiki_movies_df['Box office'].dropna()
print(box_office)

0          $21.4 million
1           $2.7 million
2            $57,718,089
3             $7,331,647
4       $6,939,946 (USA)
              ...       
7070       $19.4 million
7071       $41.9 million
7072       $76.1 million
7073       $38.4 million
7074        $5.5 million
Name: Box office, Length: 5485, dtype: object


In [44]:
## this is using a function to find out what the data types in the box office column are not strings
def is_not_a_string(x):
    return type(x) != str
box_office[box_office.map(is_not_a_string)]

34                           [US$, 4,212,828]
54      [$6,698,361 (, United States, ), [2]]
74                    [$6,488,144, (US), [1]]
126                [US$1,531,489, (domestic)]
130                          [US$, 4,803,039]
                        ...                  
6980               [$99.6, million, [4], [5]]
6994                   [$365.6, million, [1]]
6995                         [$53.8, million]
7015                     [$435, million, [7]]
7048                   [$529.3, million, [4]]
Name: Box office, Length: 135, dtype: object

In [45]:
## instead of writing out that multi-line function, we can write a lambda function to find the data types that are not
## strings
box_office[box_office.map(lambda x: type(x) != str)]

34                           [US$, 4,212,828]
54      [$6,698,361 (, United States, ), [2]]
74                    [$6,488,144, (US), [1]]
126                [US$1,531,489, (domestic)]
130                          [US$, 4,803,039]
                        ...                  
6980               [$99.6, million, [4], [5]]
6994                   [$365.6, million, [1]]
6995                         [$53.8, million]
7015                     [$435, million, [7]]
7048                   [$529.3, million, [4]]
Name: Box office, Length: 135, dtype: object

In [46]:
## the join function is kind of like concat
## we are using a space to concat the string objects
box_office = box_office.apply(lambda x: ' '.join(x) if type(x) == list else x)

#### Goal: we want to write a regex for the box office data
some of the data is written as \$123.4 million (or billion) and others are written 123,456,789
build a regular expression for each form then see what forms are left over

#### Create the first form
for the first form, our pattern match string will include six elements in the following order:
1. a dollar sign 
2. an aritrary (but non-zero) number of digits
3. an optional decimal point
4. an arbitrary (but possibly zero) number of more digits
5. a spece (maybe more than one)
6. the word "million" or "billion"

#### step 1: start with a dollar sign
regex that matches with dollar sign: "\$"

#### step 2: add an arbitrary (but non-zero) number of digits
add the "\d" character to specify digits only and the "+" modifier to capture one or more digits
<b> "\$\d+" </b>

#### step 3: add an optional decimal point
remember, the decimal point is a special character, so it needs to be escaped with a backslash
since the decimal point is optional, add a question mark after
<b> "\$\d+\.?" </b>

#### step 4: add an arbitrary (but possibly zero) number of more digits
once again, use the "\d" character to specify digits only, but with the 'asterik' modifier
this is because there may be no more digits after the decimal point
<b> "\$\d+\.?\d*"

#### step 5: add a space (maybe more than one)
use the "\s" character to match whitespace characters
<b> "\$\d+\.?\d*\s*" </b>

#### step 6: add the word "million" or "billion"
<b> "\$\d+\.?\d*\s*[mb]illion" </b>

In [50]:
## create the regex. start it with an r
form_one = r'\$\d+\.?\d*\s*[mb]illion'

In [51]:
## count up how many values match the first form (aka form_one)
box_office.str.contains(form_one, flags=re.IGNORECASE, na=False).sum()

3896

#### create the second form
will include the following elements:
1. a dollar sign
2. a group of one to three digits
3. at least one group starting with a comma and followed by exactly three digits

#### step 1: start with a dollar sign
escape the dollar sign for it to match
<b> "\$" </b>

#### step 2: add a group of one to three digits
use the "\d" character for DIGITS but modify it with curly brackets to only match one through three repetitions
<b> "\$\d{1,3}" </b>

#### step 3: match at least one group starting with a comma followed by exactly three digits
* use the string ",\d{3}"
* match the repetition of the string and then put plus sign after it "(,\d{3})+"
* insert a question mark and a colon in the beginning, this will eliminate unwanted warning messages in jupyter notebook
* finished regex <b> "\$\d{1,3}(?:,\d{3})+" </b>

In [52]:
## create form_two (look for box office amounts in number format)
form_two = r'\$\d{1,3}(?:,\d{3})+'
box_office.str.contains(form_two, flags=re.IGNORECASE, na=False).sum()

1544

In [53]:
## create boolean series to see if there are any that match both forms
matches_form_one = box_office.str.contains(form_one, flags=re.IGNORECASE, na=False)
matches_form_two = box_office.str.contains(form_two, flags=re.IGNORECASE, na=False)

In [54]:
## the squiggly line shows similar to "not"
box_office[~matches_form_one & ~matches_form_two]

34                         US$ 4,212,828
79                              $335.000
110                   $4.35-4.37 million
130                        US$ 4,803,039
600                           $5000 (US)
731                         $ 11,146,270
957                             $ 50,004
1070                          35,254,617
1147    $ 407,618 (U.S.) (sub-total) [1]
1446                        $ 11,829,959
1480                          £3 million
1611                            $520.000
1865                        ¥1.1 billion
2032                                 N/A
2091                                $309
2130               US$ 171.8 million [9]
2257                   US$ 3,395,581 [1]
2263            $ 1,223,034 ( domestic )
2347                            $282.175
2638            $ 104,883 (US sub-total)
2665         926,423 admissions (France)
2697      $ 1.7 million (US) (sub-total)
2823                            $414.000
2924                            $621.000
3088           $

In [55]:
## some values have spaces in between the dollar sign and the number
## add \s* after the dollar sign
form_one = r'\$\s*\d+\.?\d*\s*[mb]illion'
form_two = r'\$\s*\d{1,3}(?:,\d{3})+'

In [56]:
## some values use a period as a thousands separator, not a comma
form_two = r'\$\s*\d{1,3}(?:[,\.]\d{3})+(?!\s[mb]illion)'

In [57]:
## some values are given as a range
## search for any string that starts with a dollar sign and ends with a hyphen and replace
## with a dollar sign
box_office = box_office.str.replace(r'\$.*[-—–](?![a-z])', '$', regex=True)

In [58]:
## million is sometimes misspelled as millon
form_one = r'\$\s*\d+\.?\d*\s*[mb]illi?on'


In [61]:
def parse_dollars(s):
    # if s is not a string, return NaN
    if type(s) != str:
        return np.nan

    # if input is of the form $###.# million
    if re.match(r'\$\s*\d+\.?\d*\s*milli?on', s, flags=re.IGNORECASE):

        # remove dollar sign and " million"
        s = re.sub('\$|\s|[a-zA-Z]','', s)

        # convert to float and multiply by a million
        value = float(s) * 10**6

        # return value
        return value

    # if input is of the form $###.# billion
    elif re.match(r'\$\s*\d+\.?\d*\s*billi?on', s, flags=re.IGNORECASE):

        # remove dollar sign and " billion"
        s = re.sub('\$|\s|[a-zA-Z]','', s)

        # convert to float and multiply by a billion
        value = float(s) * 10**9

        # return value
        return value

    # if input is of the form $###,###,###
    elif re.match(r'\$\s*\d{1,3}(?:[,\.]\d{3})+(?!\s[mb]illion)', s, flags=re.IGNORECASE):

        # remove dollar sign and commas
        s = re.sub('\$|,','', s)

        # convert to float
        value = float(s)

        # return value
        return value

    # otherwise, return NaN
    else:
        return np.nan

In [65]:
## extract the values from box_office using str.extract
wiki_movies_df['box_office'] = box_office.str.extract(f'({form_one}|{form_two})', flags=re.IGNORECASE)[0].apply(parse_dollars)

In [66]:
wiki_movies_df['box_office']

0       21400000.0
1        2700000.0
2       57718089.0
3        7331647.0
4        6939946.0
           ...    
7071    41900000.0
7072    76100000.0
7073    38400000.0
7074     5500000.0
7075           NaN
Name: box_office, Length: 7033, dtype: float64

In [67]:
## drop the box office column because we don't need it anymore
wiki_movies_df.drop('Box office', axis=1, inplace=True)

### now time to parse budget data

In [68]:
## create a budget variable that drops null
budget = wiki_movies_df['Budget'].dropna()

In [69]:
## convert any lists to strings using lambda
budget = budget.map(lambda x: ' '.join(x) if type(x) == list else x)

In [70]:
## remove any values between a dollar sign and a hyphen
budget = budget.str.replace(r'\$.*[-—–](?![a-z])', '$', regex=True)

In [71]:
## use the same pattern to parse the box office data. apply them without modifications to the budget data
matches_form_one = budget.str.contains(form_one, flags=re.IGNORECASE, na=False)
matches_form_two = budget.str.contains(form_two, flags=re.IGNORECASE, na=False)
budget[~matches_form_one & ~matches_form_two]

136                         Unknown
204     60 million Norwegian Kroner
478                         Unknown
973             $34 [3] [4] million
1126               $120 [4] million
1226                        Unknown
1278                            HBO
1374                     £6,000,000
1397                     13 million
1480                   £2.8 million
1734                   CAD2,000,000
1913     PHP 85 million (estimated)
1948                    102,888,900
1953                   3,500,000 DM
1973                     ₤2,300,874
2281                     $14 milion
2451                     ₤6,350,000
3144                   € 40 million
3360               $150 [6] million
3418                        $218.32
3802                   £4.2 million
3906                            N/A
3959                    760,000 USD
4470                       19 crore
4641                    £17 million
5034              $$200 [4] million
5055           $155 [2] [3] million
5419                $40 [4] 

In [72]:
## remove the citation references
budget = budget.str.replace(r'\[\d+\]\s*', '')
budget[~matches_form_one & ~matches_form_two]

/var/folders/51/k9x9w__92gj0xk5c1cfb7d980000gn/T/ipykernel_87994/2849321636.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  budget = budget.str.replace(r'\[\d+\]\s*', '')


136                         Unknown
204     60 million Norwegian Kroner
478                         Unknown
973                     $34 million
1126                   $120 million
1226                        Unknown
1278                            HBO
1374                     £6,000,000
1397                     13 million
1480                   £2.8 million
1734                   CAD2,000,000
1913     PHP 85 million (estimated)
1948                    102,888,900
1953                   3,500,000 DM
1973                     ₤2,300,874
2281                     $14 milion
2451                     ₤6,350,000
3144                   € 40 million
3360                   $150 million
3418                        $218.32
3802                   £4.2 million
3906                            N/A
3959                    760,000 USD
4470                       19 crore
4641                    £17 million
5034                  $$200 million
5055                   $155 million
5419                    $40 

In [73]:
## copy the code used to parse the box office values
wiki_movies_df['budget'] = budget.str.extract(f'({form_one}|{form_two})', flags=re.IGNORECASE)[0].apply(parse_dollars)

In [74]:
## drop the original budget column
wiki_movies_df.drop('Budget', axis=1, inplace=True)

### parse release date

In [75]:
release_date = wiki_movies_df['Release date'].dropna().apply(lambda x: ' '.join(x) if type(x) == list else x)

In [76]:
## Full month name, one- to two-digit day, four-digit year (i.e., January 1, 2000)
date_form_one = r'(?:January|February|March|April|May|June|July|August|September|October|November|December)\s[123]?\d,\s\d{4}'

## Four-digit year, two-digit month, two-digit day, with any separator (i.e., 2000-01-01)
date_form_two = r'\d{4}.[01]\d.[0123]\d'

## Full month name, four-digit year (i.e., January 2000)
date_form_three = r'(?:January|February|March|April|May|June|July|August|September|October|November|December)\s\d{4}'

## Four-digit year
date_form_four = r'\d{4}'

In [77]:
## extract the dates
release_date.str.extract(f'({date_form_one}|{date_form_two}|{date_form_three}|{date_form_four})', flags=re.IGNORECASE)

0
0         July 11, 1990
1          May 17, 1990
2       August 10, 1990
3     December 25, 1990
4     December 19, 1990
...                 ...
7071  December 25, 2018
7072  December 11, 2018
7073   November 8, 2018
7074    August 31, 2018
7075      December 2018

[7001 rows x 1 columns]

In [78]:
## for dates, use built in to_datetime
wiki_movies_df['release_date'] = pd.to_datetime(release_date.str.extract(f'({date_form_one}|{date_form_two}|{date_form_three}|{date_form_four})')[0], infer_datetime_format=True)

In [79]:
## make a variable that holds the non-null values of release date in the dataframe
running_time = wiki_movies_df['Running time'].dropna().apply(lambda x: ' '.join(x) if type(x) == list else x)

In [85]:
len(running_time)

6894

In [82]:
## see how many running times look exactly like "100 mins"
running_time.str.contains(r'^\d*\s*minutes$', flags=re.IGNORECASE, na=False).sum()

6528

In [83]:
## what do the other 366 entries look like?
running_time[running_time.str.contains(r'^\d*\s*minutes$', flags=re.IGNORECASE, na=False) != True]

9                                                 102 min
26                                                 93 min
28                                                32 min.
34                                                101 min
35                                                 97 min
                              ...                        
6500       114 minutes [1] 120 minutes (extended edition)
6643                                             104 mins
6709    90 minutes (theatrical) [1] 91 minutes (unrate...
7057    108 minutes (Original cut) 98 minutes (UK cut)...
7075                Variable; 90 minutes for default path
Name: Running time, Length: 366, dtype: object

In [84]:
## look up the diff variations of "mins" by searching for strings that start with m
running_time.str.contains(r'^\d*\s*m', flags=re.IGNORECASE, na=False).sum()

6877

In [86]:
## what do the remaining 17 entries look like?
running_time[running_time.str.contains(r'^\d*\s*m', flags=re.IGNORECASE, na=False) != True]

668                     UK:84 min (DVD version) US:86 min
727                         78-102 min (depending on cut)
840                       Varies (79 [3] –84 [1] minutes)
1347                                              25 : 03
1443    United States: 77 minutes Argentina: 94 minute...
1499                                            1hr 35min
1551                                               varies
1774                    Netherlands:96 min, Canada:95 min
1777                                       approx. 14 min
2273                                           1 h 43 min
2993                                               1h 48m
3925                                              4 hours
4425    US domestic version: 86 minutes Original versi...
4967    Theatrical cut: 97 minutes Unrated cut: 107 mi...
5424                    115 [1] /123 [2] /128 [3] minutes
5447                                    1 hour 32 minutes
7075                Variable; 90 minutes for default path
Name: Running 

In [87]:
## Start with one or more digits.

## Have an optional space after the digit and before the letter "h."

## Capture all the possible abbreviations of "hour(s)." To do this, we'll make every letter in "hours" optional except the "h."

## Have an optional space after the "hours" marker.

## Have an optional number of digits for minutes.

### As a pattern, this looks like "\d+\s*ho?u?r?s?\s*\d*"

In [88]:
## add capture groups around the \d instances as well as add an alternating character
running_time_extract = running_time.str.extract(r'(\d+)\s*ho?u?r?s?\s*(\d*)|(\d+)\s*m')

In [89]:
## convert strings to numeric values. also include something that will catch empty strings
## set errors arguement to coerce - coercing the erros will turn the empty strings to a NaN
## fill the NaN to zeros using fillna()
running_time_extract = running_time_extract.apply(lambda col: pd.to_numeric(col, errors='coerce')).fillna(0)

In [90]:
## use a lambda function to import the correct format to the wiki_movies_df
wiki_movies_df['running_time'] = running_time_extract.apply(lambda row: row[0]*60 + row[1] if row[2] == 0 else row[2], axis=1)

In [91]:
## drop the running time from the dataset 
wiki_movies_df.drop('Running time', axis=1, inplace=True)